src/python/pants/core/util_rules/environments.py

```diff
@@ -363,5 +319,130 @@ def extract_process_config_from_environment(tgt: EnvironmentTarget) -> ProcessCo
     return ProcessConfigFromEnvironment(docker_image=docker_image)
 
 
+logger = logging.getLogger(__name__)
+
+
+class EnvironmentSensitiveOptionFieldMixin(ABC):
+    subsystem: type[Subsystem]
+    option_name: str
+
+
+_rules_for_subsystems: dict[type[Subsystem], set[UnionRule]] = {}
+
+
+def add_option_fields_for(subsystem: type[Subsystem]) -> Iterable[UnionRule]:
+    """Register environment fields for the `environment_sensitive` options of `subsystem`
+
+    This should be called in the `rules()` method in the file where `subsystem` is defined. It will
+    register the relevant fields under the `local_environment` and `docker_environment` targets.
+    """
+    if subsystem in _rules_for_subsystems:
+        return _rules_for_subsystems[subsystem]
```

Comment:
Is this meant to be an optimization so we don't repeat work? Instead, you could mark the function `@memoized`

Reply:
`@memoized` works at a coarser-grained level than I wanted here.